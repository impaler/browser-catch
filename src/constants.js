export const PHANTOMJS = 'phantomjs'
export const CHROME = 'chrome'

export const DRIVER_TYPES = {
  PHANTOMJS,
  CHROME
}

export const BROWSER_LOG_LEVELS = {
  WARNING: 'WARNING',
  INFO: 'INFO',
  SEVERE: 'SEVERE'
}

export const BABEL_CONFIG = {
  babelrc: false,
  presets: [
    'es2017'
  ],
  plugins: [
    'transform-es2015-modules-commonjs',
    'transform-es2015-spread'
  ]
}

export const DEFAULT_OPTIONS = {
  json: null,
  verbose: false,
  waitForExist: false,
  waitForExistReverse: false,
  waitForExistMs: 10000,
  pause: 0,
  run: null,
  webdriverHost: '127.0.0.1',
  webdriverPort: 4444,
  driverType: 'phantomjs',
  concurrent: 1
}

export const CONFIG_TYPE = {
  url: 'url',
  config: 'config',
}
