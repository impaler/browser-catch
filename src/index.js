import loadDriver from './drivers/driver'
import { create, getBrowserLogs, gotoUrl } from './client/webdriverio-log'
export { create, getBrowserLogs, gotoUrl } from './client/webdriverio-log'
const seriesSettled = require('promise-sequences').seriesSettled
const fs = require('fs')
const babelCore = require('babel-core')

const BABEL_CONFIG = {
  babelrc: false,
  presets: [
    "es2017"
  ],
  plugins: [
    "transform-es2015-modules-commonjs",
    "transform-es2015-spread"
  ]
}

export const DEFAULT_OPTIONS = {
  json: null,
  verbose: false,
  waitForExist: false,
  waitForExistReverse: false,
  waitForExistMs: 10000,
  pause: 0,
  run: null,
  webdriverHost: '127.0.0.1',
  webdriverPort: 4444,
  driverType: 'phantomjs',
  concurrent: 1
}

export const TYPE = {
  url: 'url',
  config: 'config',
}

export async function browserCatch (task, options) {
  if (task.type === TYPE.url) {
    return browserCatchUrl(task.url, options)
  } else if (task.type === TYPE.config) {
    return browserCatchConfig(task.path, options)
  } else {
    throw new Error(`browserCatch unknown task TYPE, please specify a type such as ${Object.keys(TYPE)}`)
  }
}

export async function browserCatchConfig (configPath, options) {
  options = assignDefaultOptions(options)

  let config = await readFile(configPath)
  let results

  if (config && config.urls) {
    let tasks = config.urls.map(url => () => browserCatchUrl(url, options))
    results = await seriesSettled(tasks, options.concurrent)

    let errors = results.filter(result => result.state !== 'resolved')
    if (errors.length > 0) {
      if (options.verbose) {
        errors.forEach(error => console.error(error))
      }
      throw new Error(`There were ${errors.length} errors in the task`)
    }

    return results
  } else {
    throw new Error('browserCatchConfig you need to provide a config["urls"] array in your config object')
  }
}

export async function browserCatchUrl (url, options) {
  options = assignDefaultOptions(options)
  let driver

  try {
    if (options.verbose) console.log(`Creating webdriver client`)

    driver = await loadDriver(options)
    let client = create(driver.webdriverOptions, options)
    await client.init()

    if (options.verbose) console.log(`Webdriver client created using .gotoUrl ${url}`)

    let start = Date.now()
    await gotoUrl(client, url)

    if (options.run) {
      let script = await loadScript(options.run, client, options)
      await runScript(script, client, options)
    }

    // http://webdriver.io/api/utility/waitForExist.html
    if (options.waitForExist) {
      if (options.verbose) console.log(`
Now waitForExist selector ${options.waitForExist} 
for ${options.waitForExistMs}ms & reverse ${options.waitForExistReverse}
`)
      await client.waitForExist(
        options.waitForExist,
        options.waitForExistMs,
        options.waitForExistReverse
      )
    }

    if (options.pause) {
      if (options.verbose) console.log(`Starting to pause for ${options.pause}ms`)
      await client.pause(options.pause)
    }

    let errors = await getBrowserLogs(client, options)

    let end = Date.now()
    await client.end()
    driver.kill()

    return {
      start,
      end,
      url,
      driverType: options.driverType,
      errors,
      options,
    }
  } catch (error) {
    if (driver) driver.kill()

    throw error
  }
}

export function assignDefaultOptions (options) {
  let assigned = Object.assign({}, DEFAULT_OPTIONS, options)
  if (assigned.verbose) console.log(assigned)
  return assigned
}

async function runScript (customScript, client, options) {
  try {
    await customScript(client, options)
  } catch (error) {
    if (options.verbose) console.error(`custom script has thrown an error`, error)
    throw error
  }
}

async function loadScript (scriptPath, client, options) {
  if (options.verbose) console.log(`Loading run script from path ${options.run}`)

  let customScript

  try {
    customScript = await readFileBabel(scriptPath)
  } catch (error) {
    console.error(`Could not load script your --run option ${scriptPath}`)
    throw error
  }

  return customScript
}

async function readFile (filePath) {
  if (/.js$/.test(filePath)) {
    return readFileBabel(filePath)
  } else if (/.json$/.test(filePath)) {
    return require(filePath)
  }
}

async function readFileBabel (filePath) {
  return new Promise(onReadFileBabel)

  function onReadFileBabel (resolve, reject) {
    fs.readFile(filePath, 'utf8', onReadFile)

    function onReadFile (error, fileContent) {
      if (error) reject(error)

      let fileCode = babelCore.transform(fileContent, BABEL_CONFIG)
      let file = fileCode.code
      let fileContext = requireFromString(file)
      resolve(fileContext.default)
    }
  }
}

async function requireToString (npm) {
  return new Promise(onRequireToString)

  function onRequireToString (resolve, reject) {
    const promisePolyfillPath = require.resolve(npm)
    fs.readFile(promisePolyfillPath, 'utf8', onReadFile)

    function onReadFile (error, fileContent) {
      if (error) reject(error)

      resolve(fileContent)
    }
  }
}

function requireFromString (src, filename) {
  filename = filename || '';

  var Module = module.constructor
  // console.error(filename)
  var m = new Module(filename, module.parent)
  m._compile(src, filename)
  return m.exports;
}
