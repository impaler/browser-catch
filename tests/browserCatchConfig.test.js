import test from 'ava'

import fixtureServer from './fixtures/server/server'
import { browserCatchConfig } from '../src/index'
import { stripResults } from './helpers/result'
import { writeTempJSFile, writeTempJSONFile } from './helpers/io'

const DRIVER_TYPE = 'phantomjs'

test('running a config in es6 with an array of urls', async t => {
  const serverSettings = await fixtureServer()
  const configPath = await writeTempJSFile({
    urls: [
      `${serverSettings.url}/error`,
      `${serverSettings.url}/throw`,
      `${serverSettings.url}/throw`,
      `${serverSettings.url}/error`,
    ]
  }, 'urls-config-test.js')

  const urlsResults = await browserCatchConfig(configPath)
  t.is(urlsResults.errorCount, 4)

  for (let result of urlsResults.results) {
    t.is(result.driverType, DRIVER_TYPE)
    t.is(result.errors.length, 1)
  }

  const snapshot = stripResults(urlsResults.results, serverSettings.port)
  t.snapshot(snapshot)
})

test('throwing if any of the urls fail', async t => {
  const serverSettings = await fixtureServer()
  const configPath = await writeTempJSFile({
    urls: [
      `${serverSettings.url}/error`,
      `http://somerandom/invalid`,
      `${serverSettings.url}/throw`,
      `http://yourkidding`,
      `http://doesnotexist/wow`,
    ]
  }, 'urls-config-test.js')

  const error = await t.throws(browserCatchConfig(configPath))
  t.snapshot(error)
})

test('running a config in json with an array of urls', async t => {
  const serverSettings = await fixtureServer()
  const configPath = await writeTempJSONFile({
    urls: [
      `${serverSettings.url}/error`,
      `${serverSettings.url}/throw`,
      `${serverSettings.url}/throw`,
      `${serverSettings.url}/error`,
    ]
  }, 'urls-config-test.json')

  const urlsResults = await browserCatchConfig(configPath)
  t.is(urlsResults.errorCount, 4)

  for (let result of urlsResults.results) {
    t.is(result.driverType, DRIVER_TYPE)
    t.is(result.errors.length, 1)
  }

  const snapshot = stripResults(urlsResults.results, serverSettings.port)
  t.snapshot(snapshot)
})
