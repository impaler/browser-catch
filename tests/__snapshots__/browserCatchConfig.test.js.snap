{
  "running a config in es6 with an array of urls": {
    "rejected": [],
    "resolved": [
      {
        "driverType": "phantomjs",
        "errors": [
          {
            "level": "SEVERE",
            "message": "A regular console.error\n  error (:0)"
          }
        ],
        "options": {
          "json": null,
          "verbose": false,
          "waitForExist": false,
          "waitForExistReverse": false,
          "waitForExistMs": 10000,
          "pause": 0,
          "run": null,
          "webdriverHost": "127.0.0.1",
          "webdriverPort": 4444,
          "driverType": "phantomjs",
          "concurrent": 1
        }
      },
      {
        "driverType": "phantomjs",
        "errors": [
          {
            "level": "SEVERE",
            "message": "Error: A throw in js\n  global code (http://127.0.1.1:9999/throw:6)"
          }
        ],
        "options": {
          "json": null,
          "verbose": false,
          "waitForExist": false,
          "waitForExistReverse": false,
          "waitForExistMs": 10000,
          "pause": 0,
          "run": null,
          "webdriverHost": "127.0.0.1",
          "webdriverPort": 4444,
          "driverType": "phantomjs",
          "concurrent": 1
        }
      },
      {
        "driverType": "phantomjs",
        "errors": [
          {
            "level": "SEVERE",
            "message": "Error: A throw in js\n  global code (http://127.0.1.1:9999/throw:6)"
          }
        ],
        "options": {
          "json": null,
          "verbose": false,
          "waitForExist": false,
          "waitForExistReverse": false,
          "waitForExistMs": 10000,
          "pause": 0,
          "run": null,
          "webdriverHost": "127.0.0.1",
          "webdriverPort": 4444,
          "driverType": "phantomjs",
          "concurrent": 1
        }
      },
      {
        "driverType": "phantomjs",
        "errors": [
          {
            "level": "SEVERE",
            "message": "A regular console.error\n  error (:0)"
          }
        ],
        "options": {
          "json": null,
          "verbose": false,
          "waitForExist": false,
          "waitForExistReverse": false,
          "waitForExistMs": 10000,
          "pause": 0,
          "run": null,
          "webdriverHost": "127.0.0.1",
          "webdriverPort": 4444,
          "driverType": "phantomjs",
          "concurrent": 1
        }
      }
    ]
  },
  "running a config in json with an array of urls": {
    "rejected": [],
    "resolved": [
      {
        "driverType": "phantomjs",
        "errors": [
          {
            "level": "SEVERE",
            "message": "A regular console.error\n  error (:0)"
          }
        ],
        "options": {
          "json": null,
          "verbose": false,
          "waitForExist": false,
          "waitForExistReverse": false,
          "waitForExistMs": 10000,
          "pause": 0,
          "run": null,
          "webdriverHost": "127.0.0.1",
          "webdriverPort": 4444,
          "driverType": "phantomjs",
          "concurrent": 1
        }
      },
      {
        "driverType": "phantomjs",
        "errors": [
          {
            "level": "SEVERE",
            "message": "Error: A throw in js\n  global code (http://127.0.1.1:9999/throw:6)"
          }
        ],
        "options": {
          "json": null,
          "verbose": false,
          "waitForExist": false,
          "waitForExistReverse": false,
          "waitForExistMs": 10000,
          "pause": 0,
          "run": null,
          "webdriverHost": "127.0.0.1",
          "webdriverPort": 4444,
          "driverType": "phantomjs",
          "concurrent": 1
        }
      },
      {
        "driverType": "phantomjs",
        "errors": [
          {
            "level": "SEVERE",
            "message": "Error: A throw in js\n  global code (http://127.0.1.1:9999/throw:6)"
          }
        ],
        "options": {
          "json": null,
          "verbose": false,
          "waitForExist": false,
          "waitForExistReverse": false,
          "waitForExistMs": 10000,
          "pause": 0,
          "run": null,
          "webdriverHost": "127.0.0.1",
          "webdriverPort": 4444,
          "driverType": "phantomjs",
          "concurrent": 1
        }
      },
      {
        "driverType": "phantomjs",
        "errors": [
          {
            "level": "SEVERE",
            "message": "A regular console.error\n  error (:0)"
          }
        ],
        "options": {
          "json": null,
          "verbose": false,
          "waitForExist": false,
          "waitForExistReverse": false,
          "waitForExistMs": 10000,
          "pause": 0,
          "run": null,
          "webdriverHost": "127.0.0.1",
          "webdriverPort": 4444,
          "driverType": "phantomjs",
          "concurrent": 1
        }
      }
    ]
  },
  "running a config in es6 with an array of urls rejects": {
    "rejected": [
      {
        "message": "\nThe url \"http://somerandom/invalid\" did not load, currentUrl is \"about:blank\"\n"
      },
      {
        "message": "\nThe url \"http://yourkidding\" did not load, currentUrl is \"about:blank\"\n"
      },
      {
        "message": "\nThe url \"http://doesnotexist/wow\" did not load, currentUrl is \"about:blank\"\n"
      }
    ],
    "resolved": []
  },
  "throwing if any of the urls fail": {
    "rejected": [
      {
        "message": "\nThe url \"http://somerandom/invalid\" did not load, currentUrl is \"about:blank\"\n"
      },
      {
        "message": "\nThe url \"http://yourkidding\" did not load, currentUrl is \"about:blank\"\n"
      },
      {
        "message": "\nThe url \"http://doesnotexist/wow\" did not load, currentUrl is \"about:blank\"\n"
      }
    ],
    "resolved": []
  }
}